src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -362,19 +368,44 @@ undistribute_table(PG_FUNCTION_ARGS)
 
 	SPI_connect();
 
-	List *tableDDLEvents = GetTableDDLEvents(relationId, true);
-	char *createTableQuery = linitial(tableDDLEvents);
-	list_delete_first(tableDDLEvents);
+	List *tableCommands = GetTableDDLEventsExceptCreation(relationId);
+	List *createTableCommands = GetTableCreationCommands(relationId, true);
 
-	Node *parseTree = ParseTreeNode(createTableQuery);
 
-	((CreateStmt *) parseTree)->relation->relname = "tmp1";
+	StringInfo tempName = makeStringInfo();
+	CreateTemporaryTableName(&tempName, relationName);
+
+	char *createTableQuery = NULL;
+	foreach_ptr(createTableQuery, createTableCommands)
+	{
+		Node *parseTree = ParseTreeNode(createTableQuery);
+
+		if (nodeTag(parseTree) == T_CreateStmt)
+		{
+			((CreateStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else if (nodeTag(parseTree) == T_AlterOwnerStmt)
+		{
+			((AlterOwnerStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else if (nodeTag(parseTree) == T_AlterTableStmt)
+		{
+			((AlterTableStmt *) parseTree)->relation->relname = tempName->data;
+		}
+		else
+		{
+			continue;
+		}
```

Comment:
This seems a bit fragile. I think it's better to do this in `GetTableCreationCommands`, instead of modifying the commands after they come out of there. It currently also doesn't execute any of the commands that reach the `continue` statement. That might be problematic.